Add transparent address gap limit handling & general address rotation functionality.

nuttycom opened the pull request.

nuttycom commented on zcash_client_sqlite/src/wallet/init/migrations/fix_bad_change_flagging.rs:
This could be better handled by taking a golden testing approach: branching from the commit where this test was added, create a wallet database file containing the buggy state, and store it in the repository; then, add an ordinary test that begins from that buggy database file and ensure that migrating to the current state results in the bug having been fixed.

str4d commented on zcash_client_sqlite/src/wallet/db.rs:
```diff
@@ -63,85 +61,52 @@ pub(super) const INDEX_ACCOUNTS_UIVK: &str =
+///   transparent address index. This is used for gap limit handling and is always populated if the
+///   diversifier index is in that range; since the diversifier index is stored as a byte array we
+///   cannot use SQL integer operations on it and thus need it as an integer as well.
+/// - `cached_transparent_receiver_address`: the transparent receiver component of address (which
+///   may be the same as `address` in the case of an internal-scope transparent change address or a
+///   ZIP 320 interstitial address). It is cached directly in the table to make account lookups for
+///   transparent outputs more efficient, enabling joins to [`TABLE_TRANSPARENT_RECEIVED_OUTPUTS`].
```
This documentation is now AFAICT incorrect. It is not "the transparent receiver component of address" because there are cases where `address` does not contain a transparent receiver, and yet we still populate this column.

Assuming that this is guaranteed to always be the transparent receiver derived from the same diversifier index as `address`:
```suggestion
/// - `cached_transparent_receiver_address`: the transparent address derived from the same
///   viewing key and at the same diversifier index as `address`. This may be the same as
///   `address` in the case of an internal-scope transparent change address or a
///   ZIP 320 interstitial address, and it may be a receiver within `address` in the case of a
///   Unified Address with transparent receiver. It is cached directly in the table to make account lookups for
///   transparent outputs more efficient, enabling joins to [`TABLE_TRANSPARENT_RECEIVED_OUTPUTS`].
```

str4d commented on zcash_client_sqlite/src/wallet/db.rs:
```diff
@@ -63,85 +61,52 @@ pub(super) const INDEX_ACCOUNTS_UIVK: &str =
+///   only external-key scoped addresses should be stored in this table; for purely transparent
+///   addresses, this may be an internal-scope (change) address, so that we can provide
+///   compatibility with HD-derived change addresses produced by transparent-only wallets.
+/// - `transparent_child_index`: the diversifier index, if it is in the range of a non-hardened
+///   transparent address index. This is used for gap limit handling and is always populated if the
+///   diversifier index is in that range; since the diversifier index is stored as a byte array we
+///   cannot use SQL integer operations on it and thus need it as an integer as well.
```
```suggestion
/// - `transparent_child_index`: the diversifier index in integer form, if it is in the range of a  `u31`
///   (i.e. a non-hardened transparent address index). It is used for gap limit handling, and is set
///   whenever a transparent address at a given index should be scanned at receive time. This
///   includes:
///   - Unified Addresses with transparent receivers (at any valid index).
///   - Unified Addresses without transparent receivers, but within the gap limit of potential
///     sequential transparent addresses.
///   - Transparent change addresses.
///   - ZIP 320 ephemeral addresses.
///   This column exists because the diversifier index is stored as a byte array, meaning that we
///   cannot use SQL integer operations on it for gap limit calculations, and thus need it as an
///   integer as well.
```

str4d commented on zcash_client_sqlite/src/wallet/db.rs:
```diff
@@ -63,85 +61,57 @@ pub(super) const INDEX_ACCOUNTS_UIVK: &str =
+///   may be the same as `address` in the case of an internal-scope transparent change address or a
+///   ZIP 320 interstitial address). It is cached directly in the table to make account lookups for
+///   transparent outputs more efficient, enabling joins to [`TABLE_TRANSPARENT_RECEIVED_OUTPUTS`].
+/// - `exposed_at_height`: The chain tip height at the time that the address was generated by an
+///   explicit request by the user or reserved for use in a ZIP 320 transaction. In the case of an
+///   address with its first use discovered in a transaction obtained by scanning the chain, this
+///   will be set to the mined height of that transaction.
```
```suggestion
/// - `exposed_at_height`: Our best knowledge as to when this address was first exposed to
///   the wider ecosystem. For user-generated addresses, this is the chain tip height at the
///   time that the address was generated by an explicit request by the user or reserved for
///   use in a ZIP 320 transaction. In the case of an
///   address with its first use discovered in a transaction obtained by scanning the chain, this
///   will be set to the mined height of that transaction.
```

str4d commented on zcash_client_sqlite/src/wallet/init/migrations/transparent_gap_limit_handling.rs:
```diff
@@ -0,0 +1,492 @@
+                SELECT account_id, accounts.uivk AS uivk, diversifier_index_be
+                FROM addresses
+                JOIN accounts ON accounts.id = account_id
+                GROUP BY account_id, uivk, diversifier_index_be
```
This should already be distinct, because all existing addresses were external, and `(account_id, diversifier_index_be)` was already constrained to be unique.

str4d commented on zcash_client_sqlite/src/wallet/init/migrations/transparent_gap_limit_handling.rs:
```diff
@@ -0,0 +1,492 @@
+                            cached_transparent_receiver_address = :t_addr
+                        WHERE account_id = :account_id
+                        AND diversifier_index_be = :diversifier_index_be
+                        AND key_scope = :external_scope_code
```
Not technically required as we just added this column so everything has external scope.

str4d commented on zcash_client_sqlite/src/wallet/init/migrations/transparent_gap_limit_handling.rs:
```diff
@@ -0,0 +1,492 @@
+                let diversifier_index = DiversifierIndex::from(
+                    u32::try_from(transparent_child_index).map_err(|_| {
+                        WalletMigrationError::CorruptedData(
+                            "ephermeral address indices must be in the range of `u32`".to_owned(),
```
Maybe go via `NonHardenedChildIndex` before `DiversifierIndex` to detect any "`u32` but not `u31`" corruption.

str4d commented on zcash_client_sqlite/src/wallet/db.rs:
```diff
@@ -216,6 +186,7 @@ CREATE TABLE "sapling_received_notes" (
     memo BLOB,
     commitment_tree_position INTEGER,
     recipient_key_scope INTEGER,
+    address_id INTEGER REFERENCES addresses(id),
```
Document this new column (and its nullability).

str4d commented on zcash_client_sqlite/src/wallet/db.rs:
```diff
@@ -269,6 +240,7 @@ CREATE TABLE orchard_received_notes (
     memo BLOB,
     commitment_tree_position INTEGER,
     recipient_key_scope INTEGER,
+    address_id INTEGER REFERENCES addresses(id),
```
Document this new column (and its nullability).

str4d commented on zcash_client_sqlite/src/wallet/db.rs:
```diff
@@ -339,6 +311,7 @@ CREATE TABLE transparent_received_outputs (
     script BLOB NOT NULL,
     value_zat INTEGER NOT NULL,
     max_observed_unspent_height INTEGER,
+    address_id INTEGER REFERENCES addresses(id),
```
Document this new column (and the reason it was left nullable).

str4d commented on zcash_client_sqlite/src/wallet.rs:
```diff
@@ -656,47 +751,99 @@ pub(crate) fn get_current_address<P: consensus::Parameters>(
+            :exposed_at_height
+        )
+        ON CONFLICT (account_id, diversifier_index_be, key_scope) DO UPDATE
+        SET exposed_at_height = :exposed_at_height
```
This should be the minimum of the existing and new heights, with `NULLS` greatest (i.e. overwrite a `NULL` in either direction, and take the minimum if both are non-null).

str4d commented on zcash_client_sqlite/src/wallet.rs:
```diff
@@ -656,47 +751,99 @@ pub(crate) fn get_current_address<P: consensus::Parameters>(
+            :cached_transparent_receiver_address,
+            :exposed_at_height
+        )
+        ON CONFLICT (account_id, diversifier_index_be, key_scope) DO UPDATE
```
In the case that this upsert is triggered by a note that was received in a pool that the conflicting row's address does not contain, we should upgrade the row to contain that receiver (similar to what we do in `zcashd`), because this is evidence of that receiver being exposed to the ecosystem. It's not technically evidence of _other_ receivers in `address` being exposed, so we might only want to add the one receiver triggering the upsert to the conflicting row.

nuttycom replied:
Opened #1722

str4d commented on zcash_client_sqlite/src/wallet/transparent.rs:
```diff
@@ -183,6 +190,335 @@ pub(crate) fn get_legacy_transparent_address<P: consensus::Parameters>(
+            WITH offsets AS (
+                SELECT
+                    a.transparent_child_index,
+                    LEAD(a.transparent_child_index)
```
Check what SQLite version introduced this syntax.

nuttycom replied:
This was introduced in 3.25.0

str4d commented on zcash_client_sqlite/src/wallet/transparent.rs:
```diff
@@ -183,6 +190,335 @@ pub(crate) fn get_legacy_transparent_address<P: consensus::Parameters>(
+                transparent_child_index + 1,
+                next_child_index - transparent_child_index - 1 AS gap_len
+            FROM offsets
+            WHERE gap_len >= :gap_limit OR next_child_index IS NULL
```
Add a comment here clarifying that if `gap_len` is equal to `gap_limit` (and by extension, or greater than), then `next_child_index` would not be detected by gap-limit-following scanning logic, and thus we've found the start of the corresponding gap.

str4d commented on zcash_client_sqlite/src/wallet/transparent.rs:
```diff
@@ -183,6 +190,335 @@ pub(crate) fn get_legacy_transparent_address<P: consensus::Parameters>(
+            )
+            SELECT
+                transparent_child_index + 1,
+                next_child_index - transparent_child_index - 1 AS gap_len
```
Add a comment here clarifying that both `next_child_index` and `transparent_child_index` are used, so the unused gap between them is one less than their separation.

nuttycom commented:
[force-pushed](https://github.com/zcash/librustzcash/compare/cb5adbb325ead86d812629842bd37d3d7b8c1a39..915a8d99af0c7ca19d3a2bb2c2a78612f932a1d6) to address comments from code review
then
[force-pushed](https://github.com/zcash/librustzcash/compare/915a8d99af0c7ca19d3a2bb2c2a78612f932a1d6..a0d984d6d3d52f8b6f3b0e6b83808f6e8a822ef4) to rebase on `main` to address conflicts.

daira commented on zcash_client_backend/src/data_api.rs:
````diff
@@ -1393,7 +1391,7 @@ pub trait WalletRead {
     /// This is equivalent to (but may be implemented more efficiently than):
     /// ```compile_fail
     /// Ok(
-    ///     if let Some(result) = self.get_transparent_receivers(account)?.get(address) {
````
I think this can now be equivalent to just `Ok(self.get_transparent_receivers(account, true, true)?.get(address))`. (The purpose of the `else` branch was only to include ephemeral addresses, and setting `include_ephemeral` to true does that. We also include change addresses now, but that's correct. It could be clarified by saying "a given transparent receiver, which can also be an internal or ephemeral address, ...".)

nuttycom replied:
Ephemeral address support has been removed from `get_transparent_receivers` at @str4d's request, so this still needs the if/else.

daira commented on zcash_client_backend/src/data_api.rs:
```diff
@@ -1369,6 +1365,8 @@ pub trait WalletRead {
     fn get_transparent_receivers(
         &self,
         _account: Self::AccountId,
+        _include_change: bool,
```
Do we create, or can we reasonably anticipate creating, internal *transparent* addresses other than for change? I think auto-shielding and note management will not do so.

nuttycom replied:
We will generate these as part of gap limit handling in order to make it possible to recover the full history for transparent-only wallets that follow the Bitcoin conventions for generation of change addresses.

daira replied:
I think the intent of the question was missed. It was: is it always correct to refer to these internal addresses as "change addresses", or are there / will there be non-change internal addresses (as there are in the shielded case)?

nuttycom replied:
By definition these are addresses at the `change` child index.

daira commented on zcash_client_backend/src/data_api.rs:
```diff
@@ -1414,7 +1412,10 @@ pub trait WalletRead {
     ) -> Result<Option<TransparentAddressMetadata>, Self::Error> {
         // This should be overridden.
         Ok(
-            if let Some(result) = self.get_transparent_receivers(account)?.get(address) {
```
Change the default implementation if we change the doc.

daira commented on zcash_client_backend/src/data_api.rs:
```diff
@@ -2372,6 +2374,24 @@ pub trait WalletWrite: WalletRead {
+    /// be generated using all of the available receivers for the account's UFVK.
+    ///
+    /// In the case that the diversifier index is outside of the range of valid transparent address
+    /// indexes, no transparent receiver should be generated in the resulting unified address. If a
```
```suggestion
    /// indices, no transparent receiver should be generated in the resulting unified address. If a
```
(We use "indices" far more often than "indexes"; the latter is only used a couple of times in `zcash_transparent/src/keys.rs`.)

daira commented on zcash_client_backend/src/data_api/testing/pool.rs:
```diff
@@ -598,6 +609,12 @@ pub fn send_multi_step_proposed_transfer<T: ShieldedPoolTester, DSF>(
+            st.scan_cached_blocks(h, 1);
+        }
+
         // Check that there are sent outputs with the correct values.
```
Previously there were sent outputs even before the transactions had been mined; is that no longer the case? It was intended to check it.

nuttycom replied:
The lower end of the gap (and therefore the upper end of the gap) cannot be advanced until the transaction is actually mined. Specifically, the tested behavior that is deleted where the `reserve_next_n_ephemeral_addresses` call advances the start of the gap was testing a bug.

daira replied:
The previous code did not advance the start of the gap on calling `reserve_next_n_ephemeral_addresses`. The gap was defined by `first_unmined_index` in `wallet::transparent::ephemeral::first_unsafe_index`, which comes from a query that explicitly checks `t.mined_height IS NOT NULL`:

https://github.com/zcash/librustzcash/blob/3ea34e8add1648c3dc2d56ffb26ded8cb3c587c9/zcash_client_sqlite/src/wallet/transparent/ephemeral.rs#L77-L98

nuttycom replied:
I don't think that the distinction between "sent" and "mined" is a useful one here: what we need to know is whether or not an address has been exposed, and that's why this is now modeled directly.

daira replied:
I think we might be talking past each other? You said:

> The lower end of the gap (and therefore the upper end of the gap) cannot be advanced until the transaction is actually mined.

There are two notions of gap here. The gap I'm referring to, and the comments here are referring to, is the N indices after the last one we know about (either because we generated it or because we saw it on-chain). That gives a conservative estimate of the upper bound for the range that must be scanned on recovery (for the time being; it will be extended if we see more addresses).

You're defining it to be the N indices following the last index that is known to have been mined. Both definitions are consistent with BIP 44, because that BIP *does not specify* what wallets should do when generating and using addresses. It only specifies what they should do for recovery.

The previous code didn't have the bug that you're talking about &mdash; look at the code I referred to above, which determines the "first unsafe index", i.e. the first one we cannot allocate. Maybe the previous approach was overcomplicated (well, it certainly was!), but it wasn't incorrect because it advanced the "first unsafe index" too far.

nuttycom replied:
You're correct that imprecise use of the term of "gap" is causing problems here. Specifically, I interpreted these these lines to be referring to the universally-applicable gap (i.e., the one having a lower bound determined by the last address that has been used in a mined transaction):

https://github.com/zcash/librustzcash/pull/1673/files#diff-1e9367e5fa59a041939850dc9e410a4b3f563ce75b10e2c9ae8be45b67dc19edL849-L850

daira commented on zcash_client_backend/src/data_api/testing/pool.rs:
```diff
@@ -821,85 +842,20 @@ pub fn send_multi_step_proposed_transfer<T: ShieldedPoolTester, DSF>(
-
-    // Now mine the transaction with the ephemeral output at index 10.
-    let tx = build_result.transaction();
-    let tx_index = 1;
-    let (h, _) = st.generate_next_block_from_tx(tx_index, tx);
-    st.scan_cached_blocks(h, 1);
-
-    // The above `scan_cached_blocks` does not detect `tx` as interesting to the
-    // wallet. If a transaction is in the database with a null `mined_height`,
-    // as in this case, its `mined_height` will remain null unless either
-    // `put_tx_meta` or `set_transaction_status` is called on it. The former
-    // is normally called internally via `put_blocks` as a result of scanning,
-    // but not for the case of a fully transparent transaction. The latter is
-    // called by the wallet implementation in response to processing the
-    // `transaction_data_requests` queue.
-
-    // The reservation should fail because `tx` is not yet seen as mined.
-    reservation_should_fail(&mut st, 1, 22);
-
-    // Simulate the wallet processing the `transaction_data_requests` queue.
-    let tx_data_requests = st.wallet().transaction_data_requests().unwrap();
-    assert!(tx_data_requests.contains(&TransactionDataRequest::GetStatus(tx.txid())));
-
-    // Respond to the GetStatus request.
-    st.wallet_mut()
-        .set_transaction_status(tx.txid(), TransactionStatus::Mined(h))
-        .unwrap();
-
-    // We already reserved 22 addresses, so mining the transaction with the
-    // ephemeral output at index 10 should allow 9 more (..31).
-    reservation_should_succeed(&mut st, 9);
-    reservation_should_fail(&mut st, 1, 31);
-
-    let newest_known_addrs = st
-        .wallet()
-        .get_known_ephemeral_addresses(account_id, None)
-        .unwrap();
-    assert_eq!(newest_known_addrs.len(), (GAP_LIMIT as usize) + 31);
-    assert!(newest_known_addrs.starts_with(&known_addrs));
-    assert!(newest_known_addrs[5..].starts_with(&newer_known_addrs));
```
I'm a bit concerned about loss of coverage here.

nuttycom replied:
We should look at this together; it was my conclusion that some of what was being tested here was not just implementation-dependent, but it was verifying behavior that I considered to be a bug.

daira replied:
I'm happy with this now that the `gap_limits()` test has been added.

daira commented on zcash_client_sqlite/src/error.rs:
```diff
@@ -185,12 +184,13 @@ impl fmt::Display for SqliteClientError {
-                "The proposal cannot be constructed until transactions with previously reserved ephemeral address outputs have been mined. \
-                 The ephemeral address in account {account_id:?} at index {bad_index} could not be safely reserved.",
+            SqliteClientError::ReachedGapLimit(bad_index) => write!(f,
+                "The proposal cannot be constructed until transactions with outputs to previously reserved ephemeral addresses have been mined. \
```
```suggestion
                "The proposal cannot be constructed until transactions with outputs to previously reserved transparent addresses have been mined. \
```
I think this now does not only occur for ephemeral addresses.

daira commented on zcash_client_sqlite/src/error.rs:
```diff
@@ -185,12 +184,13 @@ impl fmt::Display for SqliteClientError {
-                "The proposal cannot be constructed until transactions with previously reserved ephemeral address outputs have been mined. \
-                 The ephemeral address in account {account_id:?} at index {bad_index} could not be safely reserved.",
+            SqliteClientError::ReachedGapLimit(bad_index) => write!(f,
+                "The proposal cannot be constructed until transactions with outputs to previously reserved ephemeral addresses have been mined. \
+                 The ephemeral address at index {bad_index} could not be safely reserved.",
```
```suggestion
                "The proposal cannot be constructed until transactions with outputs to previously reserved transparent addresses have been mined. \
                 The transparent address at index {bad_index} could not be safely reserved.",
```
I think this now does not only occur for ephemeral addresses.

nuttycom replied:
The caller provided the account index for proposal construction, so we don't need to return it (and we don't have the account UUID in the place that the error is generated, and it would be a hassle to look it up.)

nuttycom replied:
`KeyScope` parameter has been added.

daira commented on zcash_client_sqlite/src/error.rs:
```diff
@@ -185,12 +184,13 @@ impl fmt::Display for SqliteClientError {
-            #[cfg(feature = "transparent-inputs")]
-            SqliteClientError::EphemeralAddressReuse(address_str, txid) => write!(f, "The ephemeral address {address_str} previously used in txid {txid} would be reused."),
+            SqliteClientError::AddressReuse(address_str, txids) => {
+                write!(f, "The address {address_str} previously used in txid(s) {:?} would be reused.", txids)
```
```suggestion
                write!(f, "The transparent address {address_str} previously used in txid(s) {:?} would be reused.", txids)
```
Also consider renaming this to `TransparentAddressReuse`.

nuttycom commented on zcash_client_backend/src/data_api/testing/pool.rs:
```diff
@@ -86,6 +86,11 @@ use zcash_protocol::PoolType;
+/// The number of ephemeral addresses that can be safely reserved without observing any
+/// of them to be mined.
+#[cfg(feature = "transparent-inputs")]
+pub const EPHEMERAL_ADDR_GAP_LIMIT: usize = 5;
```
The reason to use a smaller gap limit is that ephemeral addresses should, modulo wallet failures, *always* receive at least one (and very frequently at most one) transaction. This is the same rationale as Bitcoin wallets use for specifying a small gap limit for change addresses; I think it holds here.

daira replied:
I want to think about this carefully but can't do so today. Blocking.

nuttycom replied:
I believe that this has been discussed and resolved at this point.

daira commented on zcash_client_backend/src/sync.rs:
```diff
@@ -135,7 +135,7 @@ where
             account_id, start_height,
         );
-        refresh_utxos(params, client, db_data, account_id, start_height).await?;
+        refresh_utxos(params, client, db_data, account_id, start_height, false).await?;
```
Expand the comment to explain why it is correct to exclude ephemeral addresses.

nuttycom replied:
This sync code is in the process of being refactored by @str4d for inclusion in `zallet`; I made the minimal change to preserve the existing behavior here. This call will actually happen in a separate process that handles the retrieval queue (and will likely not even be a singular call for an account) so the goal here is minimizing disruption.

nuttycom replied:
This is no longer applicable given the changes to `get_transparent_receivers`.

daira commented on zcash_client_sqlite/src/error.rs:
```diff
@@ -119,16 +118,16 @@ pub enum SqliteClientError {
-    #[cfg(feature = "transparent-inputs")]
-    EphemeralAddressReuse(String, TxId),
+    /// The wallet attempted to create a transaction that would use of one of the wallet's
+    /// previously-used addresses, potentially creating a problem with on-chain transaction
```
```suggestion
    /// previously-used transparent addresses, potentially creating a problem with on-chain transaction
```

nuttycom replied:
I specifically generalized this to `AddressReuse` in order to make it possible to use this for unified addresses that contain a transparent receiver.

daira commented on zcash_client_sqlite/src/lib.rs:
```diff
@@ -242,13 +247,63 @@ impl fmt::Display for ReceivedNoteId {
+pub struct GapLimits {
+    external: u32,
+    transparent_internal: u32,
+    ephemeral: u32,
```
As previously commented, I think that compatibility requires we keep the ephemeral gap limit at 20. I'm also concerned about the risk of inadvertently creating similar compatibility problems for external and transparent_internal addresses, since it is incorrect to reduce any of these limits for a given wallet instance.

daira replied:
To clarify the compatibility argument: existing wallets might have gaps in their ephemeral address indices of up to 19. Sure, it's unlikely, but those wallets could have continued to allocate addresses past a gap of that size.

nuttycom replied:
I still think that this is fine, because it's even lower-probability that funds were ever _returned_ to these addresses - in order for a user to encounter a problem, they'd both have to have multiple failures _and_ then that address would have to be used for a return of funds. If after this change, a user contacts us about getting access to returned funds, _then_ I think we should consider bumping up the gap limit.

nuttycom replied:
The other thing that is relevant here is that we're making some assumptions about the generation of _all_ transparent addresses in other wallets. We only support a gap limit of 10 addresses, because we need extra space below the industry-standard of 20 addresses for handling the case where, in the future, we enable UA rotation in a way that more strongly protects against linkability. So it's (remotely) possible that some users might not find all of their funds with the current gap limit approach; I think we should just take that as a known limitation - if users really need to, they can manually fill in the gap to get to where all their addresses are visible.

daira replied:
> in order for a user to encounter a problem, they'd both have to have multiple failures and then that address would have to be used for a return of funds

I don't think that's true? To have a problem, they'd have to have had 3 or more consecutive failures to send at some point in the lifetime of the old wallet (which could just be a temporary loss of connectivity), followed by a return of funds on *any subsequent address*, and then they need to restore. They restore and they don't get the returned funds, because scanning will stop before that point.

nuttycom replied:
I think that we should have a consistent approach to "the wallet is missing funds due to the gap limit" which is: document the potential problem, and provide the user with an avenue to recover: in this case, that avenue is to send funds to the maximum-allowed ephemeral address to manually advance the gap.

daira replied:
I'm sufficiently happy with the compromise of bumping the ephemeral gap limit to 5.

str4d reviewed (commented):
Flushing some comments from an old review.

nuttycom commented on zcash_client_backend/src/data_api.rs:
```diff
@@ -1369,6 +1365,8 @@ pub trait WalletRead {
         &self,
         _account: Self::AccountId,
+        _include_change: bool,
+        _include_ephemeral: bool,
```
The outcome of this discussion will hinge upon what approach we decide upon with respect to whether or not the light wallet server is given the ability to link the ownership of transparent addresses.

The more I've thought about this, the more that I think there simply isn't a good balance between latency in discovering UTXOs and privacy to be had in this case, except by modifying the light client protocol to make it possible to do all checking for use of the wallet's transparent receivers on the client side.

nuttycom replied:
The `include_ephemeral` parameter has been removed.

str4d commented on zcash_client_sqlite/src/wallet.rs:
```diff
@@ -2536,6 +2712,8 @@ pub(crate) fn truncate_to_height<P: consensus::Parameters>(
             conn: SqlTransaction(conn),
             params: params.clone(),
+            #[cfg(feature = "transparent-inputs")]
+            gap_limits: GapLimits::default(),
```
This needs to be supplied by the caller, as `conn` and `params` were.

str4d commented on zcash_client_sqlite/src/wallet.rs:
```diff
@@ -2317,17 +2501,9 @@ pub(crate) fn store_transaction_to_be_sent<P: consensus::Parameters>(
                     },
                     None,
                     ephemeral_address,
-                    *receiving_account,
```
TODO: Determine why this was removed.

str4d commented on zcash_client_sqlite/src/wallet.rs:
```diff
@@ -2768,7 +2957,17 @@ pub(crate) fn store_decrypted_tx<P: consensus::Parameters>(
+                )?;
+
+                #[cfg(feature = "transparent-inputs")]
+                receiving_accounts.insert(_account_id, KeyScope::EXTERNAL);
```
Given that the caller provides `d_tx`, can we trust that `TransferType::Incoming` is only used when the receiving address is derieved via `KeyScope::EXTERNAL`, or is there a consistency check we need to perform here?

nuttycom replied:
It would be a bug if `TransferType` is set incorrectly, which is what this consistency check would be guarding against. I don't think that it's necessary to perform that check everywhere.

str4d commented on zcash_client_sqlite/src/wallet.rs:
```diff
@@ -2857,7 +3063,17 @@ pub(crate) fn store_decrypted_tx<P: consensus::Parameters>(
+                )?;
+
+                #[cfg(feature = "transparent-inputs")]
+                receiving_accounts.insert(_account_id, KeyScope::EXTERNAL);
```
Ditto

nuttycom replied:
See https://github.com/zcash/librustzcash/pull/1673#discussion_r1945323067

str4d commented on zcash_client_sqlite/src/wallet.rs:
```diff
@@ -2953,10 +3160,11 @@ pub(crate) fn store_decrypted_tx<P: consensus::Parameters>(
                         false,
                     )?;
 
+                    receiving_accounts.insert(account_id, key_scope);
```
Here by comparison we've fetched `key_scope` via a lookup.

nuttycom replied:
That's because we didn't do the determination elsewhere, as we do when receiving shielded notes.

str4d reviewed (commented):
Flushing partial review comments as of 4cbefae53a723b26291f09188d6e5b6bd96b8709.

nuttycom commented on zcash_client_backend/src/data_api/testing/pool.rs:
```diff
@@ -821,85 +842,20 @@ pub fn send_multi_step_proposed_transfer<T: ShieldedPoolTester, DSF>(
-    let (h, _) = st.generate_next_block_including(txids1.head);
-    st.scan_cached_blocks(h, 1);
-    reservation_should_succeed(&mut st, 2);
-    reservation_should_fail(&mut st, 1, 22);
```
This condition is tested in https://github.com/zcash/librustzcash/pull/1673/files#diff-f66631a37f9d5fcbcc2a0600364ab0735c3b2d4371118db25e7862202c0da38eR1420-R1424; I need to determine whether it's feasible to generalize this test to move it here.

str4d commented on zcash_client_sqlite/src/wallet/transparent.rs:
```diff
@@ -738,11 +1121,29 @@ pub(crate) fn put_transparent_output<P: consensus::Parameters>(
+                row.get("key_scope")?,
+            ))
+        },
+    )?;
```
It would be nice to still return `AddressNotRecognised` like the caller `put_received_transparent_utxo` did previously.

str4d reviewed (changes requested):
Flushing review comments as of f07321e9a918348110bd42aaf8d4888ac3202776. I only have `zcash_client_sqlite/src/wallet/transparent/ephemeral.rs` and `zcash_client_sqlite/src/wallet/init/migrations/transparent_gap_limit_handling.rs` left to review.

nuttycom commented on zcash_client_sqlite/src/wallet/db.rs:
```diff
@@ -1072,3 +1105,37 @@ GROUP BY
+           a.key_scope, a.diversifier_index_be, a.transparent_child_index
+    FROM orchard_received_notes orn
+    JOIN addresses a ON a.id = orn.address_id
+    JOIN transactions t ON t.id_tx = orn.tx
```
Yes; since this view is `v_address_uses`, we always need there to be a transaction. Notes are always required to be associated with a transaction.

nuttycom commented on zcash_client_sqlite/src/wallet/init/migrations/ephemeral_addresses.rs:
```diff
@@ -35,6 +44,62 @@ impl<P> schemerz::Migration<Uuid> for Migration<P> {
+                     VALUES (:account_id, :address_index, :address)",
+                )?;
+
+                // NB: we have reduced the initial space of generated ephemeral addresses
+                // from 20 addresses to 3, as ephemeral addresses should always be used in
+                // a transaction immediatly after being reserved, and as a consequence
+                // there is no significant benefit in having a larger gap limit.
```
The rationale for altering this migration is the same as the rationale for a 3-address ephemeral gap limit being acceptable in the first place: in practice, no such gaps should appear; therefore, the existing allocation of 20 addresses just means that those addresses are persisted to the wallet "early".

daira replied:
I agree with @str4d. We do not know that there are no 5-address gaps. The behaviour if there are 5-address gaps may be incorrect, but it should be deterministically incorrect, otherwise we're going to have different behaviour depending on the timing of migrations which will be impossible to debug. Changing an existing migration is fraught and should basically never be done.

nuttycom commented on zcash_client_sqlite/src/wallet/transparent.rs:
```diff
@@ -183,6 +190,390 @@ pub(crate) fn get_legacy_transparent_address<P: consensus::Parameters>(
+                ":account_id": account_id.0,
+                ":key_scope": key_scope.encode(),
+                ":gap_start": gap_start.index(),
+                ":gap_end": gap_start.saturating_add(gap_limit).index(),
```
I'm okay with that. If someone allocates 2^31 transparent addresses, their wallet will be unusable anyway.

nuttycom commented on zcash_client_sqlite/src/wallet/transparent.rs:
```diff
@@ -183,6 +190,390 @@ pub(crate) fn get_legacy_transparent_address<P: consensus::Parameters>(
+        return Err(SqliteClientError::ReachedGapLimit(
+            gap_start.index() + gap_limit,
+        ));
+    }
```
The inequality comparisons with `transparent_child_index` will already filter out rows where `transparent_child_index IS NULL`.

nuttycom commented on zcash_client_sqlite/src/wallet/transparent.rs:
```diff
@@ -183,6 +190,390 @@ pub(crate) fn get_legacy_transparent_address<P: consensus::Parameters>(
+/// The provided [`UnifiedAddressRequest`] is used to pregenerate unified addresses that correspond
+/// to the transparent address index in question; such unified addresses need not internally
+/// contain a transparent receiver, and may be overwritten when these addresses are exposed via the
+/// [`WalletWrite::get_next_available_address`] or [`WalletWrite::get_address_for_index`] methods.
```
`get_next_available_address` calls `upsert_address`, and there's actually a problem here; the current semantics of that method are that they won't alter the composition of the address, so what will happen is that the generated address won't actually be persisted (though it will be returned to the user.) What should probably happen is that some placeholder value for the address should be inserted into the `address` field, and the update should occur only if the value is that placeholder.

nuttycom replied:
The other thing here is that we actually want `get_next_available_address` to alter its behavior depending upon the address request; if the request contains a transparent element, then its diversifier index should be drawn from within the transparent range, i.e. it should be the first unreserved address within the gap; if it doesn't contain a transparent element, then it should correspond to the first unreserved diversifier index in the range of values greater than 2^32 (I presume we don't want to do anything in the space between 2^31 and 2^32 because that's the "hardened transparent" range?)

nuttycom replied:
Fixed in https://github.com/zcash/librustzcash/pull/1673/commits/d9776a4cde44324b923947ebe543b09ae8594869

nuttycom commented on zcash_client_sqlite/src/wallet/transparent.rs:
```diff
@@ -183,6 +190,390 @@ pub(crate) fn get_legacy_transparent_address<P: consensus::Parameters>(
+    };
+
+    if let Some(gap_start) = find_gap_start(conn, account_id, key_scope, gap_limit)? {
+        let range_to_store = gap_start.index()..gap_start.saturating_add(gap_limit).index();
```
I think that this is also okay.

nuttycom commented on zcash_client_sqlite/src/wallet/transparent.rs:
```diff
@@ -183,6 +190,390 @@ pub(crate) fn get_legacy_transparent_address<P: consensus::Parameters>(
+         JOIN v_received_outputs vro ON vro.transaction_id = t.id_tx
+         JOIN addresses a ON a.id = vro.address_id
+         WHERE a.address = :address
+            OR a.cached_transparent_receiver_address = :transparent_address",
```
No? if `a.cached_transparent_receiver_address` is null, that branch of the `OR` will return false, in which case `a.address = :address`, and this demonstrates address reuse.

str4d replied:
How do you get that `a.address = :address` in the example I described? I was pointing at the opposite: `:address` is a Unified Address that is not present in the `addresses` table, and has no transparent receiver. In that case:

- `a.address = :address` is false for every row.
- `:transparent_address` is `None` which maps to null in the query, so `a.cached_transparent_receiver_address = :transparent_address` is true for any row where `a.cached_transparent_receiver_address` is null, regardless of the value of `a.address`.
- Due to the `OR`, `:address` therefore results in a row being returned that should not have been a match, causing the error to be reported.

nuttycom replied:
`NULL = NULL` is false.

str4d replied:
Gah, I keep forgetting that `=` and `IS` are not isomorphic in SQL. In that case we're fine.

nuttycom commented on zcash_client_sqlite/src/wallet/transparent.rs:
```diff
@@ -183,6 +190,390 @@ pub(crate) fn get_legacy_transparent_address<P: consensus::Parameters>(
+         FROM v_address_uses au
+         JOIN addresses a ON a.id = au.address_id
+         WHERE a.account_id = :account_id
+         -- Find the maximum child index in the pre-gap range that has a mined transaction
+         -- associated with it; we can use dead reckoning here and avoid joining to the
+         -- transactions table because new gap addresses are only allocated once an
+         -- address has been used in a mined transaction.
+         AND au.transparent_child_index >= (
+             SELECT MAX((MAX(transparent_child_index) - :gap_limit), 0) AS idx
+             FROM addresses
+             WHERE key_scope = :key_scope
+         )",
```
So, for existing wallets created using `zcash_client_sqlite`, only ephemeral addresses can have a gap's worth of preallocated addresses. While `get_next_available_address` does allow generation of addresses (though it's not used in Zashi) it does so sequentially. So I think that this should be okay. I guess the alternative would be to use `find_gap_start`, and then choose the previous index; that might be a bit more robust.

nuttycom commented on zcash_client_sqlite/src/wallet/transparent.rs:
```diff
@@ -183,6 +190,390 @@ pub(crate) fn get_legacy_transparent_address<P: consensus::Parameters>(
+
+    match (h_external, h_internal) {
+        (Some(ext), Some(int)) => Ok(std::cmp::min(ext, int)),
+        (Some(ext), None) => Ok(ext),
+        (None, Some(int)) => Ok(int),
```
I didn't realize that one could use overloaded like this, that's cool.

nuttycom commented:
[force-pushed](https://github.com/zcash/librustzcash/compare/ec4c558791669ecc70f9c01211c497ec14d0233a..3bba6791e94d1665aa07d89146c61965450291e7) to address comments from code review.

nuttycom commented:
https://github.com/zcash/librustzcash/compare/f07321e..15063c6785365ccc0bb6e030e6b4ca34acb8fdb2

nuttycom commented:
[force-pushed](https://github.com/zcash/librustzcash/compare/15063c6785365ccc0bb6e030e6b4ca34acb8fdb2..fae027197f23163130840d2bd676a9ce5f05fbac) to fix bugs caught in CI

str4d commented on zcash_client_sqlite/src/error.rs:
```diff
@@ -118,17 +120,17 @@ pub enum SqliteClientError {
+    /// contains the index that could not safely be reserved.
     #[cfg(feature = "transparent-inputs")]
-    ReachedGapLimit(AccountUuid, u32),
+    ReachedGapLimit(u32, KeyScope),
```
Nit: `KeyScope` is a level above the address index, so I'd prefer the tuple being in the opposite order.

str4d commented on zcash_client_sqlite/src/wallet/transparent/ephemeral.rs:
```diff
@@ -1,184 +1,68 @@
-//! Functions for wallet support of ephemeral transparent addresses.
-use std::cmp::{max, min};
+//! Functikjns for wallet support of ephemeral transparent addresses.
```
No idea how this got broken.
```suggestion
//! Functions for wallet support of ephemeral transparent addresses.
```

str4d commented on zcash_client_sqlite/src/wallet/init/migrations/transparent_gap_limit_handling.rs:
```diff
@@ -0,0 +1,556 @@
+        conn.execute_batch(&format!(
+            r#"
+            ALTER TABLE addresses ADD COLUMN key_scope INTEGER NOT NULL DEFAULT {external_scope_code};
+            ALTER TABLE addresses ADD COLUMN transparent_child_index INTEGER;
```
An `exposed_at_height` column is not added here, so AFAICT that column will be `null` for every existing address in the table, while ephemeral addresses that get inserted have this set correctly. Given that going forward we set this for all address scopes (not just ephemeral AFAIR), we should set this correctly for existing addresses as well (to the best of our knowledge, i.e. no worse than if the wallet were recovered from seed).

nuttycom replied:
It should be safe here to set the exposed-at height to the wallet birthday; that's the most conservative approach.

nuttycom replied:
Fixed in c3e9b06fa3eef401aacafdca1dae609829bfe1d6

str4d commented on zcash_client_sqlite/src/wallet/init/migrations/transparent_gap_limit_handling.rs:
```diff
@@ -0,0 +1,556 @@
+
+        conn.execute_batch(&format!(
+            r#"
+            ALTER TABLE addresses ADD COLUMN key_scope INTEGER NOT NULL DEFAULT {external_scope_code};
```
Confirming that we did indeed only store external addresses in here previously? Definitely the case for shielded, and I believe we weren't generating change addresses for transparent.

nuttycom replied:
Correct.

str4d commented on zcash_client_sqlite/src/wallet/init/migrations/transparent_gap_limit_handling.rs:
```diff
@@ -0,0 +1,556 @@
+            CREATE TABLE addresses_new (
+                id INTEGER NOT NULL PRIMARY KEY,
+                account_id INTEGER NOT NULL,
+                key_scope INTEGER NOT NULL DEFAULT {external_scope_code},
```
Do we need to leave the default set? It's convenient for the migration (so could be retained above), but it could cause us issues during future query refactoring if the final table auto-fills the column (fixing a data error here would not be pleasant).

nuttycom replied:
Hm, I think you're correct that we may not need the default, I will check.

nuttycom replied:
Fixed in c3e9b06fa3eef401aacafdca1dae609829bfe1d6

str4d reviewed (changes requested):
Probably utACK fae027197f23163130840d2bd676a9ce5f05fbac but the `exposed_at_height` needs either addressing or reasoning.

str4d reviewed (dismissed):
utACK eca51d6b41fc4c29b86ff11bbffe892e0bd817a4

nuttycom commented:
[force-pushed](https://github.com/zcash/librustzcash/compare/eca51d6b41fc4c29b86ff11bbffe892e0bd817a4..813dd2409e62258852f1de916c439b37d896ef06) to fix conflicts with `main`

nuttycom commented:
Marked as draft at https://github.com/zcash/librustzcash/pull/1673/commits/813dd2409e62258852f1de916c439b37d896ef06 since we intend to make crate releases prior to the merge of this PR.

nuttycom commented:
[force-pushed](https://github.com/zcash/librustzcash/compare/813dd2409e62258852f1de916c439b37d896ef06..709b9be842c9788a8f7963dc1527cb1eb9abbf01) to rebase on `main`.

nuttycom commented:
[force-pushed](https://github.com/zcash/librustzcash/compare/709b9be842c9788a8f7963dc1527cb1eb9abbf01..cac7d2891dd5541891e1b4d0810d768de4d7b67c) to rebase on `main` and fix changelogs after crate releases.

nuttycom commented:
[force-pushed](https://github.com/zcash/librustzcash/compare/8db439f9378bf6d6cbabb349141083d62dc31d1b..f9fc8bbc74cf70d4314fb732c43b0a3fcf40534d) to address Clippy lints.

nuttycom commented:
Logical changes since last reviews (post-rebase-on-main) are [here](https://github.com/zcash/librustzcash/compare/cac7d28..89dcdfc4d2f1e4fffa5614f8acc7bd87945ea8bc)

nuttycom commented:
[force-pushed](https://github.com/zcash/librustzcash/compare/89dcdfc4d2f1e4fffa5614f8acc7bd87945ea8bc..4feb9f3d9a0a5e151d61a8da478e8ed2ca5e49de) to ensure no collisions can occur in timestamp-based shielded-only address generation.

nuttycom commented:
[force-pushed](https://github.com/zcash/librustzcash/compare/89dcdfc4d2f1e4fffa5614f8acc7bd87945ea8bc..d9776a4) to ensure no collisions can occur in timestamp-based shielded-only address generation.

daira commented:
Building `zcash_client_sqlite` is currently failing when the "transparent-inputs" feature is not enabled.

daira commented:
```
$ cargo +beta clippy
[...]
warning: manually reimplementing `is_power_of_two`
  --> zcash_history/src/entry.rs:35:9
   |
35 |         leaves & (leaves - 1) == 0
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider using `.is_power_of_two()`: `leaves.is_power_of_two()`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_is_power_of_two
   = note: `#[warn(clippy::manual_is_power_of_two)]` on by default

warning: `zcash_history` (lib) generated 1 warning (run `cargo clippy --fix --lib -p zcash_history` to apply 1 suggestion)
```

daira commented on zcash_client_backend/src/data_api.rs:
```diff
@@ -1229,14 +1225,19 @@ pub trait WalletRead {
     /// This will return `Ok(None)` if the account identifier does not correspond to a known
-    /// account.
-    fn get_current_address(
+    /// account, or if no previously generated address conforms to the specified request.
```
It seems undesirable to conflate those conditions. I think unknown account should always have been treated as an error here.

nuttycom replied:
Fixed in a89dd4389d2b2fd21c53b9f2c44ed920dc3bd609

daira commented on components/zcash_protocol/src/consensus.rs:
```diff
@@ -335,6 +335,16 @@ pub trait Parameters: Clone {
     }
 }
 
+impl<P: Parameters> Parameters for &P {
+    fn network_type(&self) -> NetworkType {
+        (*self).network_type()
+    }
+
+    fn activation_height(&self, nu: NetworkUpgrade) -> Option<BlockHeight> {
+        (*self).activation_height(nu)
+    }
+}
+
```
I'm confused as to why this is necessary. All crates in librustzcash compile without it, and I would have thought that receiver coercion suffices in almost all cases.

nuttycom replied:
It's not strictly necessary; I was just annoyed that I had to clone a `Parameters` instance in a case where it was really unnecessary.

daira commented on zcash_client_backend/src/data_api/testing/transparent.rs:
```diff
@@ -367,3 +370,150 @@ where
+
+    // Create a transaction & scan the block. Since the txid corresponds to one our wallet
+    // generated, this should cause the gap limit to be bumped (generating addresses with index
+    // 10..15
```
```suggestion
    // 10..15)
```

daira commented on zcash_client_backend/src/data_api/testing/pool.rs:
```diff
@@ -86,6 +86,11 @@ use zcash_protocol::PoolType;
+/// The number of ephemeral addresses that can be safely reserved without observing any
+/// of them to be mined.
+#[cfg(feature = "transparent-inputs")]
+pub const EPHEMERAL_ADDR_GAP_LIMIT: usize = 3;
```
This is quite aggressively low. The issue with a low gap limit is that there might be situations in which the wallet could request an address and then, because of some temporary condition, not be able to create or submit the transaction (i.e. it will never be mined). If it does that 3 times in succession then the wallet will be bricked for ephemeral address usage.

daira commented on zcash_client_backend/src/data_api/testing/transparent.rs:
```diff
@@ -367,3 +370,150 @@ where
+    }
+}
+
+pub fn gap_limits<DSF>(ds_factory: DSF, cache: impl TestCache, gap_limits: GapLimits)
```
This test is restoring the coverage I considered to be important. :+1:

daira commented on zcash_client_sqlite/src/error.rs:
```diff
@@ -118,17 +122,22 @@ pub enum SqliteClientError {
+    /// contains the index that could not safely be reserved.
     #[cfg(feature = "transparent-inputs")]
-    EphemeralAddressReuse(String, TxId),
+    ReachedGapLimit(KeyScope, u32),
```
How can you interpret the index without the account? To me, the index only makes sense in the context of an account, which is why the `AccountUuid` was originally included. Why is it hard to include?

nuttycom replied:
The account identifier is explicitly provided by the caller in every case where this error might be generated. The reason that the account UUID is hard to include is that this is generated in many circumstances where the account UUID is not known, because it's a few layers deep in having used the internal account ID for the lookup.

daira commented on zcash_client_sqlite/src/lib.rs:
```diff
@@ -2011,18 +2259,82 @@ mod tests {
+            .unwrap()
+            .expect("generated a shielded-only address");
+        assert_ne!(shielded_only_2, shielded_only);
+        assert!(dbg!(u128::from(di_2)) >= dbg!(u128::from(di_lower) + 10));
```
Were the `dbg!`s intended to be left in?

nuttycom replied:
No, good catch.

daira commented on zcash_client_sqlite/src/testing/pool.rs:
```diff
@@ -48,11 +48,11 @@ pub(crate) fn send_multi_step_proposed_transfer<T: ShieldedPoolTester>() {
                 e,
-                crate::error::SqliteClientError::ReachedGapLimit(acct, bad_index)
-                if acct == &account_id && bad_index == &expected_bad_index)
+                crate::error::SqliteClientError::ReachedGapLimit(_, bad_index)
+                if bad_index == &expected_bad_index)
```
```suggestion
                crate::error::SqliteClientError::ReachedGapLimit(key_scope, bad_index)
                if key_scope == KeyScope::EPHEMERAL && bad_index == &expected_bad_index)
```
(or something like that).

daira commented on zcash_client_sqlite/src/wallet.rs:
```diff
@@ -143,6 +150,15 @@ pub(crate) mod transparent;
 
+/// A constant for use in converting Unix timestamps to shielded-only diversifier indices. The
+/// value here is intended to be added to the current time, in seconds since the epoch, to obtain
+/// an index that is greater than or equal to 2^31. While it would be possible to use indices in
+/// the range 2^31..2^32, we wish to avoid any confusion with indices in the non-hardened BIP 32
+/// child index derivation space.
```
```suggestion
/// an index that is greater than or equal to 2^32. While it would be possible to use indices in
/// the range 2^31..2^32, we wish to avoid any confusion with indices in the BIP 32 child
/// index derivation space.
```

daira commented on zcash_client_sqlite/src/wallet.rs:
```diff
@@ -340,21 +366,91 @@ pub(crate) fn pool_code(pool_type: PoolType) -> i64 {
+    Ephemeral,
+}
+
+impl KeyScope {
```
Why do we need another type rather than using `zcash_primitives::legacy::keys::TransparentKeyScope`?

nuttycom replied:
This scope type applies to both shielded and transparent addresses. While the encoding is shared, I do see the zip32 shielded scope type as logically distinct from the transparent key scope type.

daira commented on zcash_client_sqlite/src/wallet.rs:
```diff
@@ -611,40 +710,206 @@ pub(crate) fn add_account<P: consensus::Parameters>(
+    // and ephemeral key scopes.
     #[cfg(feature = "transparent-inputs")]
-    transparent::ephemeral::init_account(conn, params, account_id)?;
+    for key_scope in [KeyScope::EXTERNAL, KeyScope::INTERNAL, KeyScope::Ephemeral] {
```
```suggestion
    for key_scope in [TransparentKeyScope::EXTERNAL, TransparentKeyScope::INTERNAL, TransparentKeyScope::EPHEMERAL] {
```

nuttycom replied:
Resolved via 0dcb2b2e74f6d8f0552c6cab01e2e0061150b3ce instead.

daira commented on zcash_client_sqlite/src/wallet.rs:
```diff
@@ -611,40 +710,206 @@ pub(crate) fn add_account<P: consensus::Parameters>(
+                conn,
+                params,
+                account.internal_id(),
+                KeyScope::EXTERNAL,
```
```suggestion
                TransparentKeyScope::EXTERNAL,
```

nuttycom replied:
Resolved via 0dcb2b2 instead.

daira commented on zcash_client_sqlite/src/wallet.rs:
```diff
@@ -611,40 +710,206 @@ pub(crate) fn add_account<P: consensus::Parameters>(
+                conn,
+                params,
+                account.internal_id(),
+                KeyScope::EXTERNAL,
```
```suggestion
                TransparentKeyScope::EXTERNAL,
```

nuttycom replied:
Resolved via 0dcb2b2 instead.

daira commented on zcash_client_sqlite/src/wallet.rs:
```diff
@@ -611,40 +710,206 @@ pub(crate) fn add_account<P: consensus::Parameters>(
+                        .map(|ua| (ua, j))
+                })
+                .ok_or(SqliteClientError::ReachedGapLimit(
+                    KeyScope::EXTERNAL,
```
```suggestion
                    TransparentKeyScope::EXTERNAL,
```

nuttycom replied:
Resolved via 0dcb2b2 instead.

daira commented on zcash_client_sqlite/src/error.rs:
```diff
@@ -118,17 +122,22 @@ pub enum SqliteClientError {
+    /// contains the index that could not safely be reserved.
     #[cfg(feature = "transparent-inputs")]
-    EphemeralAddressReuse(String, TxId),
+    ReachedGapLimit(KeyScope, u32),
```
```suggestion
    ReachedGapLimit(TransparentKeyScope, u32),
```

nuttycom replied:
This change was accepted as part of 0dcb2b2.

daira commented on zcash_client_sqlite/src/error.rs:
```diff
@@ -185,12 +194,25 @@ impl fmt::Display for SqliteClientError {
+                "The proposal cannot be constructed until a transaction with outputs to a previously reserved {} address has been mined. \
+                 The address at index {bad_index} could not be safely reserved.",
+                 match key_scope {
+                     KeyScope::Zip32(Scope::External) => "external transparent",
+                     KeyScope::Zip32(Scope::Internal) => "transparent change",
+                     KeyScope::Ephemeral => "ephemeral transparent",
```
```suggestion
                     TransparentKeyScope::EXTERNAL => "external transparent",
                     TransparentKeyScope::INTERNAL => "transparent change",
                     TransparentKeyScope::EPHEMERAL => "ephemeral transparent",
                     _ => unreachable!(),
```

nuttycom replied:
Included in 0dcb2b2e74f6d8f0552c6cab01e2e0061150b3ce

daira commented on zcash_client_sqlite/src/error.rs:
```diff
@@ -3,18 +3,22 @@
 use {
-    ::transparent::address::TransparentAddress, zcash_keys::encoding::TransparentCodecError,
-    zcash_primitives::transaction::TxId,
+    crate::wallet::KeyScope, ::transparent::address::TransparentAddress,
+    zcash_keys::encoding::TransparentCodecError, zip32::Scope,
```
```suggestion
    ::transparent::{address::TransparentAddress, ::keys::TransparentKeyScope},
    zcash_keys::encoding::TransparentCodecError, zip32::Scope,
```

nuttycom replied:
Included in 0dcb2b2e74f6d8f0552c6cab01e2e0061150b3ce

daira commented on zcash_client_sqlite/src/lib.rs:
```diff
@@ -111,10 +112,17 @@ use maybe_rayon::{
 };
 
+#[cfg(any(test, feature = "test-dependencies", feature = "transparent-inputs"))]
+use crate::wallet::KeyScope;
```
```suggestion
use ::transparent::keys::TransparentKeyScope;
```

nuttycom replied:
Resolved via 0dcb2b2e74f6d8f0552c6cab01e2e0061150b3ce instead.

daira commented on zcash_client_sqlite/src/lib.rs:
```diff
@@ -643,8 +776,20 @@ impl<C: Borrow<rusqlite::Connection>, P: consensus::Parameters> WalletRead for W
+        include_change: bool,
     ) -> Result<HashMap<TransparentAddress, Option<TransparentAddressMetadata>>, Self::Error> {
-        wallet::transparent::get_transparent_receivers(self.conn.borrow(), &self.params, account)
+        let key_scopes: &[KeyScope] = if include_change {
+            &[KeyScope::EXTERNAL, KeyScope::INTERNAL]
+        } else {
+            &[KeyScope::EXTERNAL]
```
```suggestion
        let key_scopes: &[TransparentKeyScope] = if include_change {
            &[TransparentKeyScope::EXTERNAL, TransparentKeyScope::INTERNAL]
        } else {
            &[TransparentKeyScope::EXTERNAL]
```

nuttycom replied:
Resolved via 0dcb2b2e74f6d8f0552c6cab01e2e0061150b3ce instead.

daira commented on zcash_client_sqlite/src/lib.rs:
```diff
@@ -756,29 +908,35 @@ impl<C: Borrow<rusqlite::Connection>, P: consensus::Parameters> WalletTest for W
+            .query_map(
+                named_params![
+                    ":txid": txid.as_ref(),
+                    ":key_scope": KeyScope::Ephemeral.encode()
```
```suggestion
                    ":key_scope": TransparentKeyScope::EPHEMERAL.encode()
```
where
```rust
impl TransparentKeyScope {
    pub fn encode(&self) -> u32 {
        self.0
    }
}
```

nuttycom replied:
Resolved via 0dcb2b2e74f6d8f0552c6cab01e2e0061150b3ce instead.

daira commented on zcash_client_sqlite/src/lib.rs:
```diff
@@ -1476,12 +1715,16 @@ impl<C: BorrowMut<rusqlite::Connection>, P: consensus::Parameters> WalletWrite f
                 wdb.conn.0,
                 &wdb.params,
                 account_id,
+                KeyScope::Ephemeral,
```
```suggestion
                TransparentKeyScope::EPHEMERAL,
```

nuttycom replied:
Resolved via 0dcb2b2e74f6d8f0552c6cab01e2e0061150b3ce instead.

daira commented on zcash_client_sqlite/src/wallet.rs:
```diff
@@ -340,21 +366,91 @@ pub(crate) fn pool_code(pool_type: PoolType) -> i64 {
+    pub(crate) const EXTERNAL: KeyScope = KeyScope::Zip32(zip32::Scope::External);
+    pub(crate) const INTERNAL: KeyScope = KeyScope::Zip32(zip32::Scope::Internal);
+
+    pub(crate) fn encode(&self) -> i64 {
+        match self {
+            KeyScope::Zip32(zip32::Scope::External) => 0i64,
+            KeyScope::Zip32(zip32::Scope::Internal) => 1i64,
+            KeyScope::Ephemeral => 2i64,
+        }
+    }
+
+    pub(crate) fn decode(code: i64) -> Result<Self, SqliteClientError> {
+        match code {
+            0i64 => Ok(KeyScope::EXTERNAL),
+            1i64 => Ok(KeyScope::INTERNAL),
+            2i64 => Ok(KeyScope::Ephemeral),
+            other => Err(SqliteClientError::CorruptedData(format!(
+                "Invalid key scope code: {}",
+                other
+            ))),
+        }
+    }
+}
+
+impl From<zip32::Scope> for KeyScope {
+    fn from(value: zip32::Scope) -> Self {
+        KeyScope::Zip32(value)
+    }
+}
+
+#[cfg(feature = "transparent-inputs")]
+impl From<KeyScope> for TransparentKeyScope {
+    fn from(value: KeyScope) -> Self {
+        match value {
+            KeyScope::Zip32(scope) => scope.into(),
+            KeyScope::Ephemeral => TransparentKeyScope::custom(2).expect("valid scope"),
+        }
     }
 }
 
```
```suggestion
```

nuttycom replied:
Resolved via 0dcb2b2e74f6d8f0552c6cab01e2e0061150b3ce instead.

daira commented on zcash_client_sqlite/src/wallet.rs:
```diff
@@ -340,21 +366,91 @@ pub(crate) fn pool_code(pool_type: PoolType) -> i64 {
-        0i64 => Some(Scope::External),
-        1i64 => Some(Scope::Internal),
-        _ => None,
+impl TryFrom<KeyScope> for zip32::Scope {
+    type Error = ();
+
+    fn try_from(value: KeyScope) -> Result<Self, Self::Error> {
+        match value {
+            KeyScope::Zip32(scope) => Ok(scope),
+            KeyScope::Ephemeral => Err(()),
+        }
     }
 }
 
```
```suggestion
```

daira commented on zcash_client_sqlite/src/wallet.rs:
```diff
@@ -611,40 +710,206 @@ pub(crate) fn add_account<P: consensus::Parameters>(
+                .query_row(
+                    named_params! {
+                        ":account_id": account.internal_id().0,
+                        ":key_scope": KeyScope::EXTERNAL.encode(),
```
```suggestion
                        ":key_scope": TransparentKeyScope::EXTERNAL.encode(),
```

daira commented on zcash_client_sqlite/src/wallet.rs:
```diff
@@ -611,40 +710,206 @@ pub(crate) fn add_account<P: consensus::Parameters>(
-            named_params![":account_uuid": account_uuid.0],
+            named_params![
+                ":account_id": account.internal_id().0,
+                ":key_scope": KeyScope::EXTERNAL.encode(),
```
```suggestion
                ":key_scope": TransparentKeyScope::EXTERNAL.encode(),
```

daira commented on zcash_client_sqlite/src/wallet.rs:
```diff
@@ -656,47 +921,171 @@ pub(crate) fn get_current_address<P: consensus::Parameters>(
+                named_params![
+                    ":account_id": account_id.0,
+                    ":diversifier_index_be": di_be,
+                    ":key_scope": KeyScope::EXTERNAL.encode(),
```
```suggestion
                    ":key_scope": TransparentKeyScope::EXTERNAL.encode(),
```

daira commented on zcash_client_sqlite/src/wallet.rs:
```diff
@@ -656,47 +921,171 @@ pub(crate) fn get_current_address<P: consensus::Parameters>(
+            ":account_id": account_id.0,
+            // the diversifier index is stored in big-endian order to allow sorting
+            ":diversifier_index_be": &di_be,
+            ":key_scope": KeyScope::EXTERNAL.encode(),
```
```suggestion
            ":key_scope": TransparentKeyScope::EXTERNAL.encode(),
```

nuttycom replied:
Resolved via 0dcb2b2e74f6d8f0552c6cab01e2e0061150b3ce instead.

daira commented on zcash_client_sqlite/src/wallet.rs:
```diff
@@ -2857,7 +3326,17 @@ pub(crate) fn store_decrypted_tx<P: consensus::Parameters>(
+                )?;
+
+                #[cfg(feature = "transparent-inputs")]
+                receiving_accounts.insert(_account_id, KeyScope::EXTERNAL);
```
```suggestion
                receiving_accounts.insert(_account_id, TransparentKeyScope::EXTERNAL);
```

nuttycom replied:
Resolved via 0dcb2b2e74f6d8f0552c6cab01e2e0061150b3ce instead.

daira commented on zcash_client_sqlite/src/wallet.rs:
```diff
@@ -3360,7 +3840,7 @@ fn flag_previously_received_change(
             named_params! {
                 ":tx": tx_ref.0,
-                ":internal_scope": scope_code(Scope::Internal)
+                ":internal_scope": KeyScope::INTERNAL.encode()
```
```suggestion
                ":internal_scope": TransparentKeyScope::INTERNAL.encode()
```

str4d commented on zcash_client_backend/CHANGELOG.md:
```diff
@@ -32,9 +58,6 @@ and this library adheres to Rust's notion of
     removed.
-- `zcash_client_backend::data_api::WalletRead::get_known_ephemeral_addresses`
-  now takes a `Range<zcash_transparent::keys::NonHardenedChildIndex>` as its
-  argument instead of a `Range<u32>`
```
When did this get into the wrong section? It's not added by this PR which means it must have been added in a PR that was included in the latest release. Are we certain this is not part of 0.17.0?

nuttycom replied:
Will check.

nuttycom replied:
The actual change was introduced in 6556505e0601a2fedaf0b43dded82d3b6c2f29a4, but the errant log entry was added in 5290d13942d4a3f149b86459ee3556d1dac196d8 - that change was factored out from this PR and I missed that the changelog contained this error.

str4d reviewed (changes requested):
Reviewed 25be66a7ea5135bee5ffbe2003b25012d3ccaee4.

daira commented on zcash_client_sqlite/src/wallet/init/migrations/fix_bad_change_flagging.rs:
```diff
@@ -52,7 +51,7 @@ impl RusqliteMigration for Migration {
                      AND {table_prefix}_received_notes.recipient_key_scope = :internal_scope"
                 ),
-                named_params! {":internal_scope": scope_code(Scope::Internal)},
+                named_params! {":internal_scope": KeyScope::INTERNAL.encode()},
```
```suggestion
                named_params! {":internal_scope": scope_code(Scope::Internal)},
```
This can't be `TransparentKeyScope::INTERNAL.encode()` because it must work when "transparent-inputs" is not enabled (also, it isn't a transparent key scope).

nuttycom replied:
Resolved via 0dcb2b2e74f6d8f0552c6cab01e2e0061150b3ce instead.

daira commented on zcash_client_sqlite/src/wallet.rs:
```diff
@@ -2768,7 +3220,17 @@ pub(crate) fn store_decrypted_tx<P: consensus::Parameters>(
+                )?;
+
+                #[cfg(feature = "transparent-inputs")]
+                receiving_accounts.insert(_account_id, KeyScope::EXTERNAL);
```
```suggestion
                receiving_accounts.insert(_account_id, TransparentKeyScope::EXTERNAL);
```

nuttycom replied:
Resolved via 0dcb2b2e74f6d8f0552c6cab01e2e0061150b3ce instead.

daira commented on zcash_client_sqlite/src/wallet/init/migrations/fix_bad_change_flagging.rs:
```diff
@@ -5,12 +5,11 @@ use std::collections::HashSet;
     wallet::{
         init::{migrations::fix_broken_commitment_trees, WalletMigrationError},
-        scope_code,
+        KeyScope,
```
```suggestion
        scope_code,
```
(or add `zip32::Scope::encode`).

nuttycom replied:
Resolved via 0dcb2b2e74f6d8f0552c6cab01e2e0061150b3ce instead.

daira commented on zcash_client_sqlite/src/wallet/init/migrations/receiving_key_scopes.rs:
```diff
@@ -109,7 +109,7 @@ impl<P: consensus::Parameters> RusqliteMigration for Migration<P> {
             &format!(
                 "ALTER TABLE sapling_received_notes ADD COLUMN recipient_key_scope INTEGER NOT NULL DEFAULT {};",
-                scope_code(Scope::External)
+                KeyScope::EXTERNAL.encode()
```
```suggestion
                scope_code(Scope::External)
```

nuttycom replied:
Resolved via 0dcb2b2e74f6d8f0552c6cab01e2e0061150b3ce instead.

daira commented on zcash_client_sqlite/src/wallet/init/migrations/receiving_key_scopes.rs:
```diff
@@ -204,7 +204,7 @@ impl<P: consensus::Parameters> RusqliteMigration for Migration<P> {
                         "UPDATE sapling_received_notes SET recipient_key_scope = :scope
                          WHERE id_note = :note_id",
-                        named_params! {":scope": scope_code(Scope::Internal), ":note_id": note_id},
+                        named_params! {":scope": KeyScope::INTERNAL.encode(), ":note_id": note_id},
```
```suggestion
                        named_params! {":scope": scope_code(Scope::Internal), ":note_id": note_id},
```

nuttycom replied:
Resolved via 0dcb2b2e74f6d8f0552c6cab01e2e0061150b3ce instead.

daira commented on zcash_client_sqlite/src/wallet/init/migrations/receiving_key_scopes.rs:
```diff
@@ -261,7 +261,7 @@ impl<P: consensus::Parameters> RusqliteMigration for Migration<P> {
                         "UPDATE sapling_received_notes SET recipient_key_scope = :scope
                          WHERE id_note = :note_id",
-                        named_params! {":scope": scope_code(Scope::Internal), ":note_id": note_id},
+                        named_params! {":scope": KeyScope::INTERNAL.encode(), ":note_id": note_id},
```
```suggestion
                        named_params! {":scope": scope_code(Scope::Internal), ":note_id": note_id},
```

nuttycom replied:
Resolved via 0dcb2b2e74f6d8f0552c6cab01e2e0061150b3ce instead.

daira commented on zcash_client_sqlite/src/wallet/init/migrations/receiving_key_scopes.rs:
```diff
@@ -604,10 +606,10 @@ mod tests {
             row_count += 1;
             let value: u64 = row.get(0).unwrap();
-            let scope = parse_scope(row.get(1).unwrap());
+            let scope = KeyScope::decode(row.get(1).unwrap()).unwrap();
```
Database consistency errors shouldn't cause panics.

daira commented on zcash_client_sqlite/src/wallet/init/migrations/receiving_key_scopes.rs:
```diff
@@ -782,10 +784,10 @@ mod tests {
             row_count += 1;
             let value: u64 = row.get(0).unwrap();
-            let scope = parse_scope(row.get(1).unwrap());
+            let scope = KeyScope::decode(row.get(1).unwrap()).unwrap();
```
Database consistency errors shouldn't cause panics.

daira commented on zcash_client_sqlite/src/wallet/init/migrations/transparent_gap_limit_handling.rs:
```diff
@@ -0,0 +1,609 @@
+
+use super::add_account_uuids;
+use crate::{
+    wallet::{self, encoding::ReceiverFlags, init::WalletMigrationError, KeyScope},
```
```suggestion
    wallet::{self, encoding::ReceiverFlags, init::WalletMigrationError},
```

nuttycom replied:
Resolved via 0dcb2b2e74f6d8f0552c6cab01e2e0061150b3ce instead.

daira commented on zcash_client_sqlite/src/wallet/init/migrations/transparent_gap_limit_handling.rs:
```diff
@@ -0,0 +1,609 @@
+use zcash_address::ZcashAddress;
+use zcash_keys::keys::UnifiedIncomingViewingKey;
+use zcash_protocol::consensus::{self, BlockHeight};
+
```
```suggestion
use zcash_transparent::keys::TransparentKeyScope;

```

nuttycom replied:
Resolved via 0dcb2b2e74f6d8f0552c6cab01e2e0061150b3ce instead.

nuttycom commented on zcash_client_sqlite/src/wallet/commitment_tree.rs:
```diff
@@ -1233,7 +1238,8 @@ mod tests {
         let data_file = NamedTempFile::new().unwrap();
-        let mut db_data = WalletDb::for_path(data_file.path(), Network::TestNetwork).unwrap();
+        let mut db_data =
+            WalletDb::for_path(data_file.path(), Network::TestNetwork, test_clock).unwrap();
```
Oh... I get it. `WalletWrite` must not be used anywhere in this test, so literally any type is fine here; it could be `()`.

nuttycom commented on zcash_client_sqlite/src/lib.rs:
```diff
@@ -960,37 +1126,46 @@ impl<C: BorrowMut<rusqlite::Connection>, P: consensus::Parameters> WalletWrite f
+            wallet::get_next_available_address(
+                wdb.conn.0,
+                &wdb.params,
+                &wdb.clock, // TODO: make an abstract Clock a field of `WalletDb`
```
Oops, leftover comment from before I did that thing!

nuttycom commented on zcash_client_sqlite/src/util.rs:
```diff
@@ -0,0 +1,63 @@
+
+use std::time::SystemTime;
+
+pub trait Clock: Clone {
```
Passing a reference to the clock where you suggested means that this bound simply isn't needed, so I'll just remove it instead.

nuttycom commented on zcash_client_sqlite/src/wallet.rs:
```diff
@@ -611,40 +710,206 @@ pub(crate) fn add_account<P: consensus::Parameters>(
+        }
+    };
+
+    let chain_tip_height = chain_tip_height(conn)?.ok_or(SqliteClientError::ChainHeightUnknown)?;
```
This doesn't make sense to me; in the case of an error, we'll have done extra unnecessary work.

nuttycom commented on zcash_client_backend/src/data_api.rs:
```diff
@@ -1229,14 +1225,19 @@ pub trait WalletRead {
-    /// account.
-    fn get_current_address(
+    /// If the `request` parameter is `None`, this will be interpreted as a request for an address
+    /// having a receiver corresponding to each data item in the account's UIVK.
```
Fixed in c1f622aaa45157b11e165733cd02f57006843331

nuttycom commented on zcash_client_sqlite/src/wallet.rs:
```diff
@@ -3106,10 +3582,14 @@ pub(crate) fn select_receiving_address<P: consensus::Parameters>(
-            let mut result = stmt.query(named_params! { ":account_uuid": account.0 })?;
+            let mut result = stmt.query(named_params! {
+                ":account_uuid": account.0,
+                ":key_scope": KeyScope::EXTERNAL.encode(),
```
Resolved via 0dcb2b2e74f6d8f0552c6cab01e2e0061150b3ce instead.

nuttycom commented on zcash_client_sqlite/src/wallet/init/migrations/receiving_key_scopes.rs:
```diff
@@ -29,7 +29,7 @@ use crate::{
         commitment_tree::SqliteShardStore,
         init::{migrations::shardtree_support, WalletMigrationError},
-        scope_code,
+        KeyScope,
```
Resolved via 0dcb2b2e74f6d8f0552c6cab01e2e0061150b3ce instead.

str4d reviewed (changes requested):
Reviewed 1199e60ae87cc156fd81c1b3b40e91dc4475e84b. Generally LGTM.

daira commented on zcash_client_backend/src/data_api.rs:
```diff
@@ -1226,18 +1226,15 @@ pub trait WalletRead {
         &self,
         account: Self::AccountId,
-        request: Option<UnifiedAddressRequest>,
+        address_filter: UnifiedAddressRequest,
```
Do you want to know the last used diversifier index, or the address with that index? In any case why not separate method(s) to get that?

nuttycom replied:
I did consider keeping the argument optional here, but on further reflection I decided that, as you say, we can have two or more different types of generation happening, and so the best option would be to make the filter required.

Ideally, we'll actually move away from the `get_current_address` type functionality entirely, in favor of using `get_next_available_address` whenever the user goes to get an address (presumably coupled to some reverse-address-book kind of functionality.)

daira commented on zcash_client_backend/src/data_api.rs:
```diff
@@ -1229,14 +1225,16 @@ pub trait WalletRead {
         account: Self::AccountId,
+        address_filter: UnifiedAddressRequest,
     ) -> Result<Option<UnifiedAddress>, Self::Error>;
 
```
```suggestion

    /// Returns the diversifier index of the most recently generated address
    /// that was allocated sequentially for this account, if the account
    /// identifier specified refers to a valid account for this wallet.
    ///
    /// This will return `Ok(None)` if no addresses have been generated
    /// using sequential allocation for this account.
    fn get_last_sequential_diversifier_index(
        &self,
        account: Self::AccountId,
    } -> Result<Option<DiversifierIndex>, Self::Error>;


    /// Returns the diversifier index of the most recently generated address
    /// obtained using time-based allocation for this account, if the account
    /// identifier specified refers to a valid account for this wallet.
    ///
    /// This will return `Ok(None)` if no addresses have been generated
    /// using time-based allocation for this account.
    fn get_last_time_based_diversifier_index(
        &self,
        account: Self::AccountId,
    } -> Result<Option<DiversifierIndex>, Self::Error>;

```

nuttycom replied:
If we have a specific use case for these methods I'm happy to add it in the future, but I don't think we should add them prospectively, and in particular, I don't think we should add them to `zcash_client_backend` as the particular approach we have chosen for address generation is specific to `zcash_client_sqlite` and may not be applicable to all `zcash_client_backend` implementations.

daira commented on zcash_client_backend/src/data_api.rs:
```diff
@@ -1229,14 +1225,16 @@ pub trait WalletRead {
-    fn get_current_address(
+    /// This will return `Ok(None)` if no previously generated address conforms to the specified
+    /// request.
+    fn get_last_generated_address(
```
```suggestion
    fn get_last_generated_address_matching(
```
I agree with this suggestion of @str4d's, independent of whether we also add `get_last_sequential_diversifier_index` and/or `get_last_time_based_diversifier_index`.

daira commented on zcash_client_sqlite/src/lib.rs:
```diff
@@ -952,45 +1134,56 @@ impl<C: BorrowMut<rusqlite::Connection>, P: consensus::Parameters> WalletWrite f
+        if let Some(account) = self.get_account(account)? {
+            if let Ok(address) = account.uivk().address(diversifier_index, request) {
+                let chain_tip_height = wallet::chain_tip_height(self.conn.borrow())?;
+                upsert_address(
```
The doc for `get_address_for_index` says (emphasis added):
> Returns `Ok(None)` in the case that it is not possible to generate an address conforming
> to the provided request at the specified diversifier index. Such a result might arise from
> either a key not being available or the diversifier index not being valid for a
> [`ReceiverRequirement::Require`]'ed receiver.
>
> Address generation should fail ***if a transparent receiver would be generated that violates***
> ***the backend's internally configured gap limit for HD-seed-based recovery***, or if an address
> has already been exposed for the given diversifier index and the given request produced an
> address having different receivers than what was originally exposed.

But I can't see where `wallet::upsert_address` enforces the highlighted part.

daira replied:
This was fixed by removing the highlighted part from the doc.

str4d reviewed (approved):
utACK d3a5bac549ceb8765407cea3a2581f9fe5d7738d

nuttycom commented:
[force-pushed](https://github.com/zcash/librustzcash/compare/d3a5bac549ceb8765407cea3a2581f9fe5d7738d..e0ae85ec8fbd27d096f373ccfca39070204a8c8f) to fix Clippy lints.

str4d reviewed (dismissed):
Re-utACK 6793a7dfe2682eadd2f58f968cbcad7f276fcd9f

str4d commented on zcash_client_backend/CHANGELOG.md:
```diff
@@ -7,6 +7,40 @@ and this library adheres to Rust's notion of
+  - Has added method `utxo_query_height` when the `transparent-inputs` feature
+    flag is active.
+  - has removed method `get_current_address`. It has been replaced by
+    added method `WalletWrite::get_last_generated_address_matching`
```
```suggestion
    added method `WalletRead::get_last_generated_address_matching`
```

daira commented on zcash_client_backend/src/data_api.rs:
```diff
@@ -2360,16 +2371,44 @@ pub trait WalletWrite: WalletRead {
+    /// diversifier index and the given request produced an address having different receivers than
+    /// what was originally exposed.
+    ///
+    /// # WARNINGS
+    /// If an address generated using this method has a transparent receiver and the
+    /// chosen diversifier index would be outside the wallet's internally-configured gap limit,
+    /// funds sent to these address are **likely to not be discovered on recovery from seed**. It
+    /// up to the caller of this method to either ensure that they only request transparent
+    /// receivers with indices within the range of a reasonable gap limit, or that they ensure that
+    /// their wallet provides backup facilities that can be used to ensure that funds sent to such
+    /// addresses are recoverable after a loss of wallet data.
```
Since the gap limit isn't documented, does a wallet have to assume that any index past the last one returned by `get_next_available_address` might give an address for which funds are not recoverable?

nuttycom replied:
`zcash_client_backend` does not specify any gap limit. A wallet application must document itself correctly for the backend in use.

daira commented on zcash_client_sqlite/src/lib.rs:
```diff
@@ -247,13 +251,124 @@ impl fmt::Display for ReceivedNoteId {
+impl GapLimits {
+    /// Constructs a new `GapLimits` value from its constituent parts.
+    ///
+    /// The following gap limits are recommended for use with this crate, and are supplied by the
+    /// [`Default`] implementation for this type.
+    /// - external addresses: 10
+    /// - transparent internal (change) addresses: 5
+    /// - ephemeral addresses: 3
```
```suggestion
    /// The gap limits recommended for use with this crate are supplied by the [`Default`]
    /// implementation for this type.
```
(avoid duplication and fix the value for ephemeral addresses which had not been updated).

daira commented on zcash_client_sqlite/src/lib.rs:
```diff
@@ -1391,13 +1391,14 @@ impl<C: BorrowMut<rusqlite::Connection>, P: consensus::Parameters, CL: Clock> Wa
                     key_scope,
                     &wdb.gap_limits,
-                    None,
+                    UnifiedAddressRequest::unsafe_custom(Allow, Allow, Require),
```
If a wallet's policy has `Deny` for either Sapling or Orchard, won't this result in `zcash_client_sqlite` implicitly generating addresses that are not compatible with the policy?

nuttycom replied:
There is not, at present, any notion of a wallet level policy; the request is made on a per-invocation basis. But in any case, this is only used at the point of _generation_, not at the point of _exposure_. The request that is used when calling `get_next_available_address` will overwrite the shielded components.

daira commented on zcash_client_sqlite/src/lib.rs:
```diff
@@ -1433,11 +1671,26 @@ impl<C: BorrowMut<rusqlite::Connection>, P: consensus::Parameters> WalletWrite f
+                account_id,
+                key_scope,
+                &wdb.gap_limits,
+                UnifiedAddressRequest::unsafe_custom(Allow, Allow, Require),
```
Same issue here as for https://github.com/zcash/librustzcash/pull/1673/files#r1978625962

daira commented on zcash_client_sqlite/src/wallet.rs:
```diff
@@ -610,41 +600,209 @@ pub(crate) fn add_account<P: consensus::Parameters>(
+                UnifiedAddressRequest::unsafe_custom(Allow, Allow, Require),
+            )?;
+
+            // Select indicies from the transparent gap limit that are available for use as
```
```suggestion
            // Select indices from the transparent gap limit that are available for use as
```

daira commented on zcash_client_sqlite/src/wallet.rs:
```diff
@@ -610,41 +600,209 @@ pub(crate) fn add_account<P: consensus::Parameters>(
+                account.internal_id(),
+                KeyScope::EXTERNAL,
+                gap_limits,
+                UnifiedAddressRequest::unsafe_custom(Allow, Allow, Require),
```
Same issue as https://github.com/zcash/librustzcash/pull/1673/files#r1978625962

daira commented on zcash_client_sqlite/src/wallet.rs:
```diff
@@ -610,41 +600,209 @@ pub(crate) fn add_account<P: consensus::Parameters>(
+                    .expect("gap limit fits in usize"),
+            )?;
+
+            // Find the first index that generates an address conforming to the request.
```
Did we resolve the issue with there being a significant probability of not finding any Sapling addresses within the gap limit?

daira commented on zcash_client_sqlite/src/wallet.rs:
```diff
@@ -610,41 +600,209 @@ pub(crate) fn add_account<P: consensus::Parameters>(
+    let require_flags = ReceiverFlags::required(requirements);
+    let omit_flags = ReceiverFlags::omitted(requirements);
+    // This returns the most recently exposed external-scope address (the address that was exposed
+    // at the greatest block height) that conforms to the specified requirements.
```
```suggestion
    // at the greatest block height, using the largest diversifier index to break ties)
    // that conforms to the specified requirements.
```

daira commented on zcash_client_sqlite/src/wallet.rs:
```diff
@@ -3028,6 +3392,19 @@ pub(crate) fn store_decrypted_tx<P: consensus::Parameters>(
+            account_id,
+            key_scope,
+            gap_limits,
+            UnifiedAddressRequest::unsafe_custom(Allow, Allow, Require),
```
Same issue here as https://github.com/zcash/librustzcash/pull/1673/files#r1978625962

daira commented on zcash_client_sqlite/src/wallet/db.rs:
```diff
@@ -63,85 +61,77 @@ pub(super) const INDEX_ACCOUNTS_UIVK: &str =
-/// - The `cached_transparent_receiver_address` column contains the transparent receiver component
-///   of the UA. It is cached directly in the table to make account lookups for transparent outputs
-///   more efficient, enabling joins to [`TABLE_TRANSPARENT_RECEIVED_OUTPUTS`].
+///   This column exists because the diversifier index is stored as a byte array, meaning that we
+///   cannot use SQL integer operations on it for gap limit calculations, and thus need it as an
+///   integer as well.
```
It occurs to me that we could store the diversifier index as pairs of SQL integers, the high 56 bits and the low 32 bits. That would be normalized, easily sortable in SQL, and easy to check in SQL that the high bits are all 0. Filed #1731.

daira commented on zcash_client_sqlite/src/wallet/db.rs:
```diff
@@ -202,6 +192,25 @@ CREATE TABLE "transactions" (
+/// - `memo`: the memo output associated with the note, if known
+/// - `commitment_tree_position`: the 0-based index of the note in the leaves of the note
+///   commitment tree.
+/// - `receipient_key_scope`: the ZIP 32 key scope of the key that decrypted this output,
```
```suggestion
/// - `recipient_key_scope`: the ZIP 32 key scope of the key that decrypted this output,
```

daira commented on zcash_client_sqlite/src/wallet/db.rs:
```diff
@@ -254,6 +264,26 @@ CREATE TABLE sapling_received_note_spends (
+/// - `memo`: the memo output associated with the note, if known
+/// - `commitment_tree_position`: the 0-based index of the note in the leaves of the note
+///   commitment tree.
+/// - `receipient_key_scope`: the ZIP 32 key scope of the key that decrypted this output,
```
```suggestion
/// - `recipient_key_scope`: the ZIP 32 key scope of the key that decrypted this output,
```

daira commented on zcash_client_sqlite/src/wallet/init/migrations.rs:
```diff
@@ -87,6 +88,8 @@ pub(super) fn all_migrations<P: consensus::Parameters + 'static>(
     //                         fix_broken_commitment_trees         add_account_uuids
     //                                     |                               |
     //                          fix_bad_change_flagging     v_transactions_additional_totals
+    //                                                                 |
+    //                                                     transparent_gap_limit_handling
```
```suggestion
    //                                                                     |
    //                                                       transparent_gap_limit_handling
```

daira commented on zcash_client_sqlite/src/wallet/init/migrations/ephemeral_addresses.rs:
```diff
@@ -35,6 +44,62 @@ impl<P> schemerz::Migration<Uuid> for Migration<P> {
+                )?;
+
+                // NB: we have reduced the initial space of generated ephemeral addresses
+                // from 20 addresses to 3, as ephemeral addresses should always be used in
+                // a transaction immediatly after being reserved, and as a consequence
```
```suggestion
                // a transaction immediately after being reserved, and as a consequence
```

daira replied:
```suggestion
                // from 20 addresses to 5, as ephemeral addresses should always be used in
                // a transaction immediately after being reserved, and as a consequence
```

daira commented on zcash_client_sqlite/src/wallet/init/migrations/transparent_gap_limit_handling.rs:
```diff
@@ -0,0 +1,610 @@
+
+        // Add foreign key references from the *_received_{notes|outputs} tables to the addresses
+        // table to make it possible to identify which address was involved. These foreign key
+        // columns  must be nullable as for shielded account-internal. Ideally the foreign key
```
```suggestion
        // columns must be nullable as for shielded account-internal. Ideally the foreign key
```

daira commented on zcash_client_sqlite/src/wallet/transparent.rs:
```diff
@@ -183,6 +191,427 @@ pub(crate) fn get_legacy_transparent_address<P: consensus::Parameters>(
+                -- so the gap between them is one less than their difference
+                next_child_index - transparent_child_index - 1 AS gap_len
+            FROM offsets
+            -- if gap_len is equal to the gap limit, then we have found a gap.
```
```suggestion
            -- if gap_len is at least the gap limit, then we have found a gap.
```

daira commented on zcash_client_sqlite/src/wallet/transparent.rs:
```diff
@@ -183,6 +191,427 @@ pub(crate) fn get_legacy_transparent_address<P: consensus::Parameters>(
+/// transparent addresses is available from the first gap in existing indices of addresses at which
+/// a received transaction has been observed on the chain, for each key scope.
+///
+/// The provided [`UnifiedAddressRequest`] is used to pregenerate unified addresses that correspond
```
```suggestion
/// The provided [`UnifiedAddressRequest`] is used to pre-generate unified addresses that correspond
```

daira commented on zcash_client_sqlite/src/wallet/transparent.rs:
```diff
@@ -949,21 +1401,59 @@ mod tests {
+                find_gap_start(&db.conn, account_id, KeyScope::Ephemeral, db.gap_limits.ephemeral()), Ok(addr_index)
+                    if addr_index == Some(NonHardenedChildIndex::ZERO)
+            );
+            //assert_matches!(ephemeral::first_unstored_index(&db.conn, account_id), Ok(addr_index) if addr_index == GAP_LIMIT);
```
Is this assertion correct? Delete it if it isn't correct, otherwise uncomment it.

daira reviewed (dismissed):
utACK

daira reviewed (approved):
Self-ACK for the last three commits.

str4d reviewed (approved):
Re-utACK 19e0e3986e047a226ac8a6be66bb263d3e1447bc
